[Merged by Bors] - Improve debugging experience for builder proposals

paulhauner opened the pull request:
## Issue Addressed

NA

## Proposed Changes

This PR sets out to improve the logging/metrics experience when interacting with the builder. Namely, it:

- Adds/changes metrics (see "Metrics Changes" section).
- Adds new logs which show the duration of requests to the builder/local EL.
- Refactors existing logs for consistency and so that the `parent_hash` is include in all relevant logs (we can grep for this field when trying to trace the flow of block production).


Additionally, when I was implementing this PR I noticed that we skip some verification of the builder payload in the scenario where the builder return `Ok` but the local EL returns with `Err`. Namely, we were skipping the bid signature and other values like parent hash and prev randao. In this PR I've changed it so we *always* check these values and reject the bid if they're incorrect. With these changes, we'll sometimes choose to skip a proposal rather than propose something invalid -- that's the only side-effect to the changes that I can see.

## Metrics Changes

- Changed: `execution_layer_request_times`:
    - `method = "get_blinded_payload_local"`: time taken to get a payload from a local EE.
    - `method = "get_blinded_payload_builder"`: time taken to get a blinded payload from a builder.
    - `method = "post_blinded_payload_builder"`: time taken to get a builder to reveal a payload they've previously supplied us.
- `execution_layer_get_payload_outcome`
    - `outcome = "success"`: we successfully produced a payload from a builder or local EE.
    - `outcome = "failure"`: we were unable to get a payload from a builder or local EE.
- New: `execution_layer_builder_reveal_payload_outcome`
    - `outcome = "success"`: a builder revealed a payload from a signed, blinded block.
    - `outcome = "failure"`: the builder did not reveal the payload.
- New: `execution_layer_get_payload_source`
    - `type = "builder"`: we used a payload from a builder to produce a block.
    - `type = "local"`: we used a payload from a local EE to produce a block.
- New: `execution_layer_get_payload_builder_rejections` has a `reason` field to describe why we rejected a payload from a builder.
- New: `execution_layer_payload_bids` tracks the bid (in gwei) from the builder or local EE (local EE not yet supported, waiting on EEs to expose the value). Can only record values that fit inside an i64 (roughly 9 million ETH).
## Additional Info

NA

paulhauner committed:
Refactor `get_blinded_payload`

paulhauner committed:
Add metrics and logs for payload unblinding

paulhauner committed:
Run `cargo fmt`

paulhauner committed:
Tidy, add buckets

paulhauner commented:
This might be a good one for @realbigsean, since he's familiar with this code? I'm keen to hear your thoughts on the extra checks for builder-only proposals :pray:

michaelsproul commented on beacon_node/execution_layer/src/lib.rs:
```diff
@@ -608,130 +611,174 @@ impl<T: EthSpec> ExecutionLayer<T> {
-                            forkchoice_update_params,
-                        )
+
+                    // Wait for the builder *and* relay to produce a payload (or return an error).
```
```suggestion
                    // Wait for the builder *and* local EL to produce a payload (or return an error).
```

michaelsproul commented on beacon_node/execution_layer/src/lib.rs:
```diff
@@ -608,130 +611,174 @@ impl<T: EthSpec> ExecutionLayer<T> {
-                                "block_hash" => ?header.block_hash(),
+                                "Received local and builder payloads";
+                                "relay_block_hash" => ?header.block_hash(),
+                                "local_block_hash" => %local.block_hash(),
```
Is the long-form for the relay block hash here intentional, as it's most likely to be selected?

paulhauner replied:
Ah yeah, I actually intended to make both of these long-form. I figured that long-form would make it easier for us to search third-party services. I find it a little frustrating when we have short-form and then need to cross-reference with something like beaconcha.in.

I've gone and changed these in 26118a49a. I ended up changing all roots/hashes in this file to be `Display`, hopefully this is not perceived as over-zealous.

michaelsproul replied:
I'm onboard with the long hashes! Agree 100% about the annoyances of short hashes.

I think you got the impls switched in https://github.com/sigp/lighthouse/commit/26118a49af162ea07c57ce00823b9ac1692c3bd1 though. The `Debug` implementation is the long-form one

michaelsproul commented on beacon_node/execution_layer/src/lib.rs:
```diff
@@ -1404,18 +1451,183 @@ impl<T: EthSpec> ExecutionLayer<T> {
+                Err(e) => crit!(
+                    self.log(),
+                    "Builder failed to reveal payload";
+                    "info" => "this relay failure has caused a missed proposal",
```
It's possible that the relay still unblinded the block and published it

```suggestion
                    "info" => "this relay failure may cause a missed proposal",
```

michaelsproul commented on beacon_node/execution_layer/src/lib.rs:
```diff
@@ -1404,18 +1451,183 @@ impl<T: EthSpec> ExecutionLayer<T> {
+async fn timed_future<F: Future<Output = T>, T>(metric: &str, future: F) -> (T, Duration) {
+    let start = Instant::now();
+    let result = future.await;
+    let duration = Instant::now().duration_since(start);
```
Equivalently:

```suggestion
    let duration = start.elapsed();
```

(https://doc.rust-lang.org/std/time/struct.Instant.html#method.elapsed)

michaelsproul commented on beacon_node/execution_layer/src/lib.rs:
```diff
@@ -608,130 +611,174 @@ impl<T: EthSpec> ExecutionLayer<T> {
+                                        "parent_hash" => %parent_hash,
                                     );
+                                    Err(Error::CannotProduceHeader)
                                 }
```
IMO we shouldn't log a crit in the case where the profit threshold isn't met, given that we expect it often and want to encourage users to set profit thresholds.

Previously we had an info log for low value payloads, maybe we could return an enum from `verify_builder_bid` like:

```rust
#[must_use]
pub enum ValidBuilderPayload {
   Satisfactory,
   LowValue {
       profit_threshold: U256,
       payload_value: U256,
   },
}
```

Or just be careful to handle the `LowValue` case separately when calling `verify_builder_bid`.

paulhauner replied:
Great point, I've fixed this in [0a86b24](https://github.com/sigp/lighthouse/pull/3725/commits/0a86b244bffe686cdcbb75d6c01e42fad08bb417).

michaelsproul reviewed (commented):
Looks great on the whole, will be very useful for trying out some new relays ;)

michaelsproul commented:
Related features that might be cool to add to this PR (or a follow-up):

- Count metrics for the number of builder proposals vs local proposals (vs misses?)
- Counts for each `reason` that a builder proposal is not made (an `IntCounterVec`).

paulhauner committed:
Apply suggestions from code review

Co-authored-by: Michael Sproul <micsproul@gmail.com>

paulhauner committed:
Print full hash/root in several places

paulhauner committed:
Handle low value payload when there's no alternative

paulhauner committed:
Add more metrics

paulhauner commented:
> Related features that might be cool to add to this PR (or a follow-up):

I added a bunch of new metrics in [339698b](https://github.com/sigp/lighthouse/pull/3725/commits/339698ba3c3c4fb5ed2ca1994ae52e96c4071541) and updated the description to describe them. Good idea.

EDIT: a fixed a bug in those metrics in [55e0b9e](https://github.com/sigp/lighthouse/pull/3725/commits/55e0b9e53a749d437ee4918d41c751a38e345166).

EDIT 2: In 54390beae I swapped to tracking the bid in gwei, since we'd be stuck at ~9 ETH when trying to fit wei inside an i64.

paulhauner committed:
Fix metric tracking for get_payload

paulhauner committed:
Log bid in gwei

michaelsproul commented on beacon_node/execution_layer/src/lib.rs:
```diff
@@ -580,6 +592,40 @@ impl<T: EthSpec> ExecutionLayer<T> {
+                );
+                metrics::inc_counter_vec(
+                    &metrics::EXECUTION_LAYER_GET_PAYLOAD_SOURCE,
+                    &[metrics::LOCAL],
```
```suggestion
                    &[metrics::BUILDER],
```

michaelsproul commented on beacon_node/execution_layer/src/lib.rs:
```diff
@@ -606,132 +652,193 @@ impl<T: EthSpec> ExecutionLayer<T> {
                         "slot" => ?slot,
                         "pubkey" => ?pubkey,
-                        "parent_hash" => ?parent_hash,
+                        "parent_hash" => %parent_hash,
```
The `%` form is the short-form: https://github.com/sigp/lighthouse/pull/3725#discussion_r1033178270

paulhauner replied:
Ah of course 🤦 Fixed in e8bd2be26

michaelsproul commented on beacon_node/execution_layer/src/lib.rs:
```diff
@@ -606,132 +652,193 @@ impl<T: EthSpec> ExecutionLayer<T> {
+                    info!(
+                        self.log(),
+                        "Requested blinded execution payload";
+                        "relay_fee_recipient" => match &relay_result {
```
We may want to use the debug impl for the fee recipients too

paulhauner replied:
Good call, also fixed in e8bd2be26

michaelsproul commented on beacon_node/execution_layer/src/lib.rs:
```diff
@@ -606,132 +652,193 @@ impl<T: EthSpec> ExecutionLayer<T> {
+                                    warn!(
                                         self.log(),
-                                        "The value offered by the connected builder does not meet \
-                                        the configured profit threshold. Using local payload.";
-                                        "configured_value" => ?configured_value, "relay_value" => ?relay_value
+                                        "Builder returned invalid payload";
+                                        "info" => "using local payload",
+                                        "reason" => %reason,
+                                        "relay_block_hash" => %header.block_hash(),
+                                        "parent_hash" => %parent_hash,
                                     );
```
I think maybe this should only be an `info` log in the case the bid threshold isn't met. Otherwise we'll get a lot of noisy warnings (better than noisy crits, but still).

paulhauner replied:
Good call, I've made it so it's only an info when the payload is valid but ignored: e0b2b1aa4

I've maintained the `warn` for invalid payloads since I'm keen to be loud about builders that are doing the wrong thing.

michaelsproul commented on beacon_node/execution_layer/src/lib.rs:
```diff
@@ -1402,20 +1510,225 @@ impl<T: EthSpec> ExecutionLayer<T> {
+                        &metrics::EXECUTION_LAYER_BUILDER_REVEAL_PAYLOAD_OUTCOME,
+                        &[metrics::FAILURE],
+                    );
+                    crit!(
```
Maybe an error seeing as this is kinda outside our control, and maybe not that severe?

```suggestion
                    error!(
```

paulhauner replied:
My understanding of real-world events so far is that if a builder isn't revealing blocks to the BN then those blocks are lost and the relay is probably bugged. Given that, I'd be tempted to start with a `crit` here and then perhaps relax it to an `error` if we see that builders are failing to reveal to us but actually publishing on the network. I think this is particularly relevant whilst we don't have tracking of missed blocks in the validator manager.

paulhauner replied:
> the relay is probably bugged

... and probably not going to do anything about it for several hours 😅

michaelsproul replied:
Yeah sounds good. I thought there'd been at least one incident where the blocks were getting broadcast but not revealed to the BN, but I don't have a source. Happy to leave it as a CRIT.

michaelsproul reviewed (commented):
Looking good!

Do we have this running on any test infra yet?

paulhauner committed:
Update beacon_node/execution_layer/src/lib.rs

Co-authored-by: Michael Sproul <micsproul@gmail.com>

paulhauner committed:
Avoid warning with low-value builder payload

paulhauner committed:
Move roots/hashes back to debug

paulhauner committed:
Merge branch 'unstable' into builder-metrics

michaelsproul commented on beacon_node/execution_layer/src/lib.rs:
```diff
@@ -608,130 +654,202 @@ impl<T: EthSpec> ExecutionLayer<T> {
+                                Err(reason) => {
+                                    metrics::inc_counter_vec(
+                                        &metrics::EXECUTION_LAYER_GET_PAYLOAD_BUILDER_REJECTIONS,
+                                        &[reason.as_ref().as_ref()],
```
Does this compile?

```suggestion
                                        &[reason.as_ref()],
```

paulhauner replied:
No, unfortunately not. The error is `Box`ed to satisfy clippy.

paulhauner replied:
`(*reason).as_ref()` works, but I recall you saying you hate that many moons ago 😅

michaelsproul replied:
Ah yeah, fair enough

michaelsproul commented on beacon_node/execution_layer/src/lib.rs:
```diff
@@ -608,130 +654,202 @@ impl<T: EthSpec> ExecutionLayer<T> {
+                                Err(reason) => {
+                                    metrics::inc_counter_vec(
+                                        &metrics::EXECUTION_LAYER_GET_PAYLOAD_BUILDER_REJECTIONS,
+                                        &[reason.as_ref().as_ref()],
```
```suggestion
                                        &[reason.as_ref()],
```

paulhauner replied:
As above.

michaelsproul reviewed (approved):
Approved!

paulhauner committed:
Fix bug with gwei calculation

paulhauner commented:
I fixed a sneaky bug here: [c6256e1](https://github.com/sigp/lighthouse/pull/3725/commits/c6256e10adfe4c57751218fdfdfc3839fdadb5a7)

paulhauner commented:
Alrighty, I'm running this on some Prater nodes to good result! Thanks for the review ☺️ 

bors r+

paulhauner commented:
@michaelsproul I just got woken up in the middle of the night because [a block](https://beaconcha.in/slot/5263648) was not revealed by a builder but was actually broadcast 😅 Perhaps you're right about dropping that log level.

michaelsproul commented:
Let's drop it to `ERROR`! That won't wake you up, right?
